lib/gc/Transforms/Microkernel/ConvertLinalgToMicrokernel.cpp

```diff
@@ -0,0 +1,366 @@
+    return failure();
+  }
+  if (contractionDims->m.size() != 1 || contractionDims->n.size() != 1 ||
+      // batch-reduce dim for BRGEMM should be identified as one of k dim
+      // including VNNI & non-VNNI cases
+      (contractionDims->k.size() != 2 && contractionDims->k.size() != 3) ||
+      !contractionDims->batch.empty()) {
+    LLVM_DEBUG(llvm::dbgs() << "[checkStructure] Wrong dimensions\n");
+    LLVM_DEBUG(llvm::dbgs()
+               << "[checkStructure] " << contractionDims->m.size() << " "
+               << contractionDims->n.size() << " " << contractionDims->k.size()
+               << " " << contractionDims->batch.size() << "\n");
+    return failure();
+  }
+  unsigned classifiedLoops =
+      contractionDims->m.size() + contractionDims->n.size() +
+      contractionDims->k.size() + contractionDims->batch.size();
+  if (linalgOp.getNumLoops() != classifiedLoops) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "[checkStructure] Not all loops are classified\n");
+    return failure();
+  }
+
+  unsigned mAffinePos = contractionDims->m[0];
+  unsigned nAffinePos = contractionDims->n[0];
+  // contractionDims.k could be of 2 cases:
+  //     1. dims.k.size() == 2: non-VNNI, K = dims.k[1]
+  //     2. dims.k.size() == 3: VNNI, K = dims.k[1] * dims.k[2]
+  unsigned batchAffinePos = contractionDims->k.front();
+  SmallVector<unsigned, 2> kAffinePos;
+  if (contractionDims->k.size() == 2)
+    kAffinePos = {contractionDims->k[1]};
+  else if (contractionDims->k.size() == 3)
+    kAffinePos = {contractionDims->k[1], contractionDims->k[2]};
+  else
+    return failure();
+
+  LLVM_DEBUG(llvm::dbgs() << "[inferBrgemmDims] Candidate dims: "
+                          << "\n");
+  LLVM_DEBUG(llvm::dbgs() << "[inferBrgemmDims] m pos in affine: " << mAffinePos
+                          << "\n");
+  LLVM_DEBUG(llvm::dbgs() << "[inferBrgemmDims] n pos in affine: " << nAffinePos
+                          << "\n");
+  for (auto kp : kAffinePos)
+    LLVM_DEBUG(llvm::dbgs()
+               << "[inferBrgemmDims] k pos in affine: " << kp << "\n");
+  LLVM_DEBUG(llvm::dbgs() << "[inferBrgemmDims] batch pos in affine: "
+                          << batchAffinePos << "\n");
+
+  OpOperand *operandA = linalgOp.getDpsInputOperands()[0];
+  OpOperand *operandB = linalgOp.getDpsInputOperands()[1];
+
+  BrgemmDims brgemmDims;
+
+  auto checkAndGetPosInDomain = [&](int64_t &dim, ArrayRef<unsigned> dimPos,
+                                    OpOperand *operand) {
+    auto pos = getPosInDomain(dimPos, operand, linalgOp);
+    assert(pos && "Cannot find position in codomain");
+    dim = *pos;
+  };
+
+  // A(batch, m, k)
+  checkAndGetPosInDomain(brgemmDims.batchDimA, batchAffinePos, operandA);
+  checkAndGetPosInDomain(brgemmDims.leadingDimA, {mAffinePos}, operandA);
+  checkAndGetPosInDomain(brgemmDims.minorDimA, kAffinePos, operandA);
+  // B(batch, k, n) or B(batch, k/vnni_step, n, vnni_step)
```

Comment:
For B leading dimension (`K`), will always use `kAffinePos[0]` to find pos in B domain regardless of `k.size()`, because generic form of vnni linalg batch GEMM access B with `B[b, k0, n, k1]`, we only need to concern about `k0`